Create a Feature Branch Approval Policy (#62)
openssl/general-policies: policies/feature-branch-approval-policy.md

# Feature Branch Approval Policy

> The [Time-based Release Policy] defines a regular twice a year schedule for new

Reviewer: ```suggestion
The [Time-based Release Policy] defines a regular twice-a-year schedule for new
```

Reviewer: It feels like it should have hyphens...

OpenSSL versions. In the case of large features that cannot reasonably be

> implemented and reviewed within a single PR there is a risk that the feature
> will only be partially implemented and merged to master by the time that a
> feature freeze occurs for the next release.

Reviewer: ```suggestion
implemented and reviewed within a single PR there is a risk that only a
partial implementation of the feature will be merged to the master branch
by the time a feature freeze occurs for the next release.
```

Reviewer 2: maybe s/only a partial/an incomplete or incorrect/

To avoid this problem feature branches can be used. A feature branch is simply
a branch in the main git repository that can be used as a target branch for
PRs related to a single new feature.

A feature branch will be created based on the current state of master at the
time of its creation. If a feature branch exists for a protracted period of time
then it will need to be rebased on a regular basis. The process for doing so is
deferred to other policies.

The creation of a new feature branch implies some overhead for the project to
manage that feature branch (such as for example periodic rebases, and to review

> PRs targetting it). Such a feature branch would typically only be created where

Reviewer: ```suggestion
PRs targeting it). Such a feature branch would typically only be created where
```

it has been agreed that the feature is desirable; exists on the [Project Board];
is of sufficiently high priority; and there are expected to be resources
available to work on the feature development and to review it. For this reason
the creation of new feature branches is controlled via an approval process.

Submitting a request for a new feature branch should be done by creating an
issue in the [openssl/project] repository. Such a request can be created by

> anyone without restriction. The request should specify:

Reviewer: ```suggestion
anyone. The request should specify:
```

Reviewer: Simplify?

> - The name of the feature branch requested

Reviewer 2: My suggestion:
```suggestion
- The name of the feature branch requested in the form of `feature/<name>`
```

This way we will use the same pattern as for the already existing `feature/quic-server` branch.

> - The name of the feature branch requested

Reviewer: Should we put here other requirements?
- Feature detailed description;
- The purpose of the feature [and how it aligns with the project's goals];
- Expected scope and impact of this feature. Any changes to existing functionality?
- Potential benefits or drawbacks;
- Dependencies;
- Plan to test? Unit tests, integration tests, and any manual testing procedures.
- Documentation;
- Responsibility for maintaining this feature once it's integrated;

Reviewer 2: This sounds like too much bureaucratic boilerplate. Also there will be an accompanying openssl/openssl feature issue linked where most of this should be described already. I.e., the openssl/project feature branch request should happen only after the feature is already discussed in the openssl/openssl feature issue and there is a consensus that doing such feature in a single PR is not reasonable.

> Responsibility for maintaining this feature once it's integrated;

That is an interesting point - I do not think we should ever accept a feature that we are not willing to maintain as the team. BTW - how we can hold the contributors to be responsible for this promised responsibility if they simply later disappear?

Author: I've not made a change here, as per @t8m's comments

Author: @arapov - my suggestion is that we move some of these things into the feature request issue template:

https://github.com/openssl/openssl/blob/master/.github/ISSUE_TEMPLATE/feature_request.md?plain=1

- A link to an issue describing the requested feature that will be developed on
  the branch
- An estimate of the expected OpenSSL release that the feature will be included

>   in

Reviewer: The requester will have difficulty determining this as it depends on our capacity to review. Should it be framed as an estimation of when the work on the feature is expected to be done? Ultimately, it is up to us which release we will accept.

- Whether there are any known resources being made available to work on the
  feature

The approval of a new feature branch is ultimately performed by the OMC. The OMC
may delegate the authority to make such approvals on its behalf to another body
at the discretion of the OMC.

Once approved by the OMC a feature branch may be created. All PRs relevant to
that feature should target that branch and go through the normal code review
process.

Once the feature is complete and all relevant PRs have been merged to the branch,

> all commits can be pushed to the master branch directly without further review.

Reviewer: ```suggestion
all commits may then be eligible for merging into the master branch, subject to
final review.
```

Reviewer: Merging without review sounds dangerous and reckless, although I understand that everything merged into the feature branch has already been reviewed and did not show issues in CI.

Nonetheless, we should conduct an extra review. It might be less thorough, but a final check sounds reasonable.

Wdyt?

Author: I've added some words about this

# Feature Branch Approval Policy

The [Time-based Release Policy] defines a regular twice-a-year schedule for new
OpenSSL versions. In the case of large features that cannot reasonably be
implemented and reviewed within a single PR there is a risk that only an
incomplete or incorrect implementation of the feature will be merged to the
master branch by the time a feature freeze occurs for the next release.

To avoid this problem feature branches can be used. A feature branch is simply
a branch in the main git repository that can be used as a target branch for
PRs related to a single new feature.

A feature branch will be created based on the current state of master at the
time of its creation. If a feature branch exists for a protracted period of time
then it will need to be rebased on a regular basis. This process should be

> performed by a committer an an as-needed basis and can be done without further

Reviewer 2: Typo: `in an`
